docs/pipelines.md

```diff
@@ -516,6 +517,203 @@ In particular:
 
 The `description` field is an optional field and can be used to provide description of the `Pipeline`.
 
+## Adding `Finally` to the `Pipeline`
+
+You can specify a list of one or more final tasks under `finally` section. Final tasks are guaranteed to be executed
+after all `PipelineTasks` under `tasks` have completed regardless of success or error. Final tasks are very similar to
+`PipelineTasks` under `tasks` section and follow the same syntax. Each final task must have a
+[valid](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names) `name` and a `taskRef` or
+`taskSpec`. For example:
+
+```yaml
+spec:
+  tasks:
+    - name: tests
+      taskRef:
+        Name: integration-test
+  finally:
+    - name: cleanup-test
+      taskRef:
+        Name: cleanup
+```
+
+### Specifying `Resources` in Final Tasks
+
+Similar to `tasks`, you can use [PipelineResources](#specifying-resources) as inputs and outputs for
+final tasks in the Pipeline. The only difference here is, final tasks with an input resource can have a `from` clause
+but that resource has to rely on output of a `PipelineTask` from `tasks` section. For example:
```

Comment:
What happens if the task we take the resource from failed?
I think we should not allow finally tasks to depend on things that may not be there, or we could allow that only if the input resource is an optional one, so that the task can still be executed even if the input is missing.

Reply:
Irrespective of whether task succeed or fails, referenced resource container does get created, lets say if its image resource, image might not have created and might not exist at the registry but from tekton perspective image resource still exists. 

How do we distinguish between failure to create and push image v/s task not creating or pushing image even after executing successfully?

Please point me to a specific scenario if I am missing anything 🙏

Reply:
> Irrespective of whether task succeed or fails, referenced resource container does get created, 

Do you mean the resource folder in the artifact storage?
The task with the resource may also be skipped because of previous failures.

> lets say if its image resource, image might not have created and might not exist at the registry but from tekton perspective image resource does exist.
> 
> How do we distinguish between failure to create and push image v/s task not creating or pushing image even after executing successfully?
> 
> Please point me to a specific scenario if I am missing anything 🙏

I think I see your point. Using from enforces ordering, which is ignored in case of `finally` tasks. The task will look for the content of the input resource in the artifact storage and it will have to be able to cope with both cases, content is there or is not.
This is actually a very good behavior I think, but we might want to add a comment in the docs here to say that the input resource may not be populated there, and so a finally task will have to handle that.

Reply:
`from` is not being supported and being tracked in https://github.com/tektoncd/pipeline/issues/2711